File: src/lib/analytics/nutrient-correlation.ts
import { pearsonCorrelation, type CorrelationResult } from './correlation';
import { benjaminiHochberg } from './stats';
import { shiftDate } from './date-utils';

export type NutrientCorrelation = {
	nutrientKey: string;
	correlation: CorrelationResult;
	/** Benjamini–Hochberg adjusted p across every nutrient screened. */
	qValue: number;
	/** Number of nutrients that were testable in this screen. */
	comparisons: number;
};

const MIN_PAIRS = 7;
const MIN_ABS_R = 0.15;
const FDR_LEVEL = 0.1;

/**
 * Screens every nutrient key against a daily outcome. The strongest of thirty
 * correlations is an upward-biased maximum even when nothing is there, so the
 * screen applies Benjamini–Hochberg control across the keys tested and only
 * returns those under the FDR threshold, most extreme first. Callers pairing a
 * nutrient with body weight should pass day-over-day weight *changes* as the
 * outcome: two trending level series correlate spuriously.
 */
export function computeNutrientOutcomeCorrelations(
	dailyNutrients: { date: string; nutrients: Record<string, number | null> }[],
	outcomes: { date: string; value: number }[],
	lagDays: number = 0
): NutrientCorrelation[] {
	const outcomeMap = new Map<string, number>();
	for (const o of outcomes) {
		outcomeMap.set(o.date, o.value);
	}

	const allKeys = new Set<string>();
	for (const day of dailyNutrients) {
		for (const key of Object.keys(day.nutrients)) {
			allKeys.add(key);
		}
	}

	const tested: { nutrientKey: string; correlation: CorrelationResult }[] = [];

	for (const key of allKeys) {
		const paired: { nutrient: number; outcome: number }[] = [];
		let nullCount = 0;

		for (const day of dailyNutrients) {
			const nutrientValue = day.nutrients[key] ?? null;
			if (nutrientValue === null || nutrientValue === undefined) {
				nullCount++;
				continue;
			}

			const outcomeDate = lagDays === 0 ? day.date : shiftDate(day.date, lagDays);
			const outcomeValue = outcomeMap.get(outcomeDate);
			if (outcomeValue === undefined) continue;

			paired.push({ nutrient: nutrientValue, outcome: outcomeValue });
		}

		const totalDays = dailyNutrients.length;
		if (totalDays === 0 || nullCount / totalDays > 0.5) continue;

		if (paired.length < MIN_PAIRS) continue;

		const xVals = paired.map((p) => p.nutrient);
		const yVals = paired.map((p) => p.outcome);

		tested.push({ nutrientKey: key, correlation: pearsonCorrelation(xVals, yVals) });
	}

	const qValues = benjaminiHochberg(tested.map((t) => t.correlation.pValue));
	const results = tested
		.map((t, i) => ({ ...t, qValue: qValues[i], comparisons: tested.length }))
		.filter((t) => Math.abs(t.correlation.r) >= MIN_ABS_R && t.qValue <= FDR_LEVEL);

	results.sort((a, b) => Math.abs(b.correlation.r) - Math.abs(a.correlation.r));

	return results;
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/analytics/NutrientCorrelation.kt
package com.bissbilanz.analytics

import kotlin.math.abs

data class NutrientCorrelation(
    val nutrientKey: String,
    val correlation: CorrelationResult,
    /** Benjamini–Hochberg adjusted p across every nutrient screened. */
    val qValue: Double,
    /** Number of nutrients that were testable in this screen. */
    val comparisons: Int,
)

private const val MIN_PAIRS = 7
private const val MIN_ABS_R = 0.15
private const val FDR_LEVEL = 0.1

/**
 * Screens every nutrient key against a daily outcome with Benjamini–Hochberg
 * control across the keys tested, returning only those under the FDR threshold,
 * most extreme first. Callers pairing a nutrient with body weight should pass
 * day-over-day weight *changes* as the outcome.
 */
fun computeNutrientOutcomeCorrelations(
    dailyNutrients: List<Pair<String, Map<String, Double?>>>,
    outcomes: List<Pair<String, Double>>,
    lagDays: Int = 0,
): List<NutrientCorrelation> {
    val outcomeMap = mutableMapOf<String, Double>()
    for ((date, value) in outcomes) outcomeMap[date] = value
    val allKeys = LinkedHashSet<String>()
    for ((_, nutrients) in dailyNutrients) allKeys.addAll(nutrients.keys)
    val tested = mutableListOf<Pair<String, CorrelationResult>>()
    for (key in allKeys) {
        val totalDays = dailyNutrients.size
        val nullCount = dailyNutrients.count { (_, nutrients) -> nutrients[key] == null }
        if (totalDays == 0 || nullCount.toDouble() / totalDays > 0.5) continue
        val xValues = mutableListOf<Double>()
        val yValues = mutableListOf<Double>()
        for ((date, nutrients) in dailyNutrients) {
            val nv = nutrients[key] ?: continue
            val outcomeDate = if (lagDays == 0) date else shiftDate(date, lagDays)
            val outcome = outcomeMap[outcomeDate] ?: continue
            xValues.add(nv)
            yValues.add(outcome)
        }
        if (xValues.size < MIN_PAIRS) continue
        tested.add(Pair(key, pearsonCorrelation(xValues.toDoubleArray(), yValues.toDoubleArray())))
    }
    val qValues = benjaminiHochberg(tested.map { it.second.pValue })
    return tested
        .mapIndexed { i, (key, corr) -> NutrientCorrelation(key, corr, qValues[i], tested.size) }
        .filter { abs(it.correlation.r) >= MIN_ABS_R && it.qValue <= FDR_LEVEL }
        .sortedByDescending { abs(it.correlation.r) }
}
